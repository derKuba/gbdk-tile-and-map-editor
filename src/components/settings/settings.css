:host {
  display: block;
  display: flex;
  flex-direction: column;
}

.settingsIcon {
  width: 30px;
  height: 30px;
  margin-left: auto;
}

.formLine {
  margin-bottom: 15px;
}

.settingsForm {
  border: 1px solid #c0c0c0;
  padding: 0 20px 0 20px;
  margin-bottom: 25px;
  display: none;
  text-align: center;
  width: 450px;
  margin: 0 auto;
  margin-bottom: 45px;
}

.settingsForm input[type="number"]{
  width: 50px;
  text-align: center;
  margin-left: 10px;
}

.settingsForm span {
  margin: 0 30px 0 30px;
}

.settingsForm.visible{
  display: block;
}