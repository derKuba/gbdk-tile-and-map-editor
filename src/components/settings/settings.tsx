/* eslint-disable max-lines */
/* eslint-disable max-lines-per-function */
import { Component, Host, h, State, Prop } from '@stencil/core';

import { Settings } from '../../models/settingsModel';
import Tunnel from '../../provider/StateProvider';

@Component({
    tag: 'mvjp-settings-ui',
    styleUrl: 'settings.css',
    shadow: true,
})
export class SettingsUI {
    /**
     *   global settings
     */
    @Prop() settings: Settings;

    /**
     *   setter for global settings
     */
    @Prop() setSettings: Function;

    @State() isContentVisible: boolean = false;

    private toggleVisibility = () => {
        this.isContentVisible = !this.isContentVisible;
    };

    private handleChange(event) {
        switch (event.target.name) {
            case 'width':
                this.settings.fieldWidth = parseInt(event.target.value, 10);
                break;
            case 'height':
                this.settings.fieldHeight = parseInt(event.target.value, 10);
                break;

            case 'spriteAmount':
                this.settings.spriteAmount = parseInt(event.target.value, 10);
                break;

            case 'grid':
                this.settings.isGridVisible = event.target.checked;
                break;

            case 'numbers':
                this.settings.isFieldNumberVisible = event.target.checked;
                break;

            default:
                break;
        }
        this.setSettings(this.settings);
    }

    render() {
        const settingsFormVisibleClass = this.isContentVisible ? 'visible' : '';
        return (
            <Host>
                <img
                    class="settingsIcon"
                    onClick={() => this.toggleVisibility()}
                    src={'./assets/icon/freepik_repair.png'}
                    alt="settings icon"
                />

                <div class={`settingsForm ${settingsFormVisibleClass}`}>
                    <h2>Settings</h2>

                    <div class="formLine">
                        <label>
                            Width:
                            <input
                                name="width"
                                type="number"
                                min="1"
                                max="99"
                                value={this.settings.fieldWidth}
                                onInput={event => this.handleChange(event)}
                            />
                        </label>
                        <span> x </span>
                        <label>
                            Height:
                            <input
                                name="height"
                                type="number"
                                min="1"
                                max="99"
                                onInput={event => this.handleChange(event)}
                                value={this.settings.fieldHeight}
                            />
                        </label>
                    </div>

                    <div class="formLine">
                        <label>
                            Number of Tiles:
                            <input
                                name="spriteAmount"
                                type="number"
                                min="1"
                                max="99"
                                onInput={event => this.handleChange(event)}
                                value={this.settings.spriteAmount}
                            />
                        </label>
                    </div>

                    <div class="formLine">
                        <label>
                            Grid:
                            <input
                                name="grid"
                                type="checkbox"
                                value="isGridVisible"
                                onInput={event => this.handleChange(event)}
                                checked={this.settings.isGridVisible}
                            />
                        </label>

                        <label>
                            Numbers:
                            <input
                                name="numbers"
                                type="checkbox"
                                value="isFieldNumberVisible"
                                onInput={event => this.handleChange(event)}
                                checked={this.settings.isFieldNumberVisible}
                            />
                        </label>
                    </div>
                </div>
            </Host>
        );
    }
}
Tunnel.injectProps(SettingsUI, ['settings', 'setSettings']); // eslint-disable-line @stencil/ban-side-effects
