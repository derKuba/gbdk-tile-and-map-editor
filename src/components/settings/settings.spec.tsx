import { newSpecPage } from '@stencil/core/dist/testing';
import { h } from '@stencil/core';

import { SettingsUI } from './settings';

describe('mvjp-settings', () => {
    let settings;
    let page;
    beforeEach(async () => {
        settings = {
            fieldWidth: 8,
            fieldHeight: 8,
            spriteAmount: 12,
            isGridVisible: false,
            isFieldNumberVisible: false,
        };

        page = await newSpecPage({
            components: [SettingsUI],
            template: () => {
                return <mvjp-settings-ui settings={settings} setSettings={() => {}} />;
            },
        });
    });

    it('builds', () => {
        expect(new SettingsUI()).toBeTruthy();
    });

    it('matches snapshot', async () => {
        const setSettings = () => jest.fn();

        const snapShotPage = await newSpecPage({
            components: [SettingsUI],
            template: () => {
                return <mvjp-settings-ui settings={settings} setSettings={setSettings} />;
            },
        });

        expect(snapShotPage.root).toMatchSnapshot();
    });

    it('should toggle the visibility', async () => {
        page.root.shadowRoot.querySelector('.settingsIcon').click();

        await page.waitForChanges();

        expect(page.rootInstance.isContentVisible).toBe(true);
    });

    it('should set fieldwith', async () => {
        const widthElement = page.root.shadowRoot.querySelector('input[name="width"]');
        const event = new Event('input', { bubbles: true, cancelable: true });

        widthElement.value = 12;
        widthElement.dispatchEvent(event);

        await page.waitForChanges();

        expect(page.rootInstance.settings.fieldWidth).toBe(12);
    });

    it('should set fieldHeight', async () => {
        const heightElement = page.root.shadowRoot.querySelector('input[name="height"]');
        const event = new Event('input', { bubbles: true, cancelable: true });

        heightElement.value = 12;
        heightElement.dispatchEvent(event);

        await page.waitForChanges();

        expect(page.rootInstance.settings.fieldHeight).toBe(12);
    });

    it('should set spriteAmount', async () => {
        const spriteAmountElement = page.root.shadowRoot.querySelector('input[name="spriteAmount"]');
        const event = new Event('input', { bubbles: true, cancelable: true });

        spriteAmountElement.value = 7;
        spriteAmountElement.dispatchEvent(event);

        await page.waitForChanges();

        expect(page.rootInstance.settings.spriteAmount).toBe(7);
    });

    it('should set grid', async () => {
        const event = new Event('input', { bubbles: true, cancelable: true });
        const grid = page.root.shadowRoot.querySelector('input[name="grid"]');
        grid.checked = true;
        grid.dispatchEvent(event);

        await page.waitForChanges();

        expect(page.rootInstance.settings.isGridVisible).toBe(true);
    });

    it('should set numbers', async () => {
        const event = new Event('input', { bubbles: true, cancelable: true });
        const grid = page.root.shadowRoot.querySelector('input[name="numbers"]');
        grid.checked = true;
        grid.dispatchEvent(event);

        await page.waitForChanges();

        expect(page.rootInstance.settings.isFieldNumberVisible).toBe(true);
    });
});
