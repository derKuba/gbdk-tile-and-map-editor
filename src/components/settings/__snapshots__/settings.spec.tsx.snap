// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mvjp-settings matches snapshot 1`] = `
<mvjp-settings-ui>
  <mock:shadow-root>
    <img alt="settings icon" class="settingsIcon" src="./assets/icon/freepik_repair.png">
    <div class="settingsForm">
      <h2>
        Settings
      </h2>
      <div class="formLine">
        <label>
          Width:
          <input max="99" min="1" name="width" type="number" value="8">
        </label>
        <span>
          x
        </span>
        <label>
          Height:
          <input max="99" min="1" name="height" type="number" value="8">
        </label>
      </div>
      <div class="formLine">
        <label>
          Number of Tiles:
          <input max="99" min="1" name="spriteAmount" type="number" value="12">
        </label>
      </div>
      <div class="formLine">
        <label>
          Grid:
          <input name="grid" type="checkbox" value="isGridVisible">
        </label>
        <label>
          Numbers:
          <input name="numbers" type="checkbox" value="isFieldNumberVisible">
        </label>
      </div>
    </div>
  </mock:shadow-root>
</mvjp-settings-ui>
`;
